jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c

```diff
@@ -1687,6 +1688,128 @@ ecma_builtin_typedarray_prototype_find_index (ecma_value_t this_arg, /**< this a
 } /* ecma_builtin_typedarray_prototype_find_index */
 
+/**
+ * The %TypedArray%.prototype object's 'indexOf' and 'lastIndexOf' routine helper
+ *
+ * See also:
+ *         ECMA-262 v6, 22.2.3.13, 22.2.3.16
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_typedarray_prototype_index_helper (ecma_value_t this_arg, /**< this argument */
+                                                const ecma_value_t args[], /**< arguments list */
+                                                ecma_length_t args_number, /**< number of arguments */
+                                                bool is_last_index_of) /**< true - lastIndexOf routine
+                                                                         false - indexOf routine */
+{
+  if (!ecma_is_typedarray (this_arg))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not a TypedArray."));
+  }
+
+  ecma_object_t *typedarray_p = ecma_get_object_from_value (this_arg);
+  uint32_t length = ecma_typedarray_get_length (typedarray_p);
+  lit_magic_string_id_t class_id = ecma_object_get_class_name (typedarray_p);
+  lit_utf8_byte_t *typedarray_buffer_p = ecma_typedarray_get_buffer (typedarray_p);
+  uint8_t shift = ecma_typedarray_get_element_size_shift (typedarray_p);
+  uint8_t element_size = (uint8_t) (1 << shift);
+  uint32_t limit = length * element_size;
+  ecma_value_t ret_value = ECMA_VALUE_EMPTY;
+  uint32_t from_index;
+
+  if (args_number == 0 || length == 0)
+  {
+    return ecma_make_integer_value (-1);
+  }
+  if (args_number == 1)
+  {
+    from_index = is_last_index_of ? length - 1 : 0;
+  }
+  else
+  {
+    ecma_number_t num_var;
+    ret_value = ecma_get_number (args[1], &num_var);
+
+    if (ECMA_IS_VALUE_ERROR (ret_value))
+    {
+      return ret_value;
+    }
+
+    from_index = ecma_builtin_helper_array_index_normalize (num_var, length, is_last_index_of);
+  }
+
+  ecma_value_t search_element = args[0];
+
+  if (ecma_number_is_nan (search_element))
+  {
+    return ecma_make_integer_value (-1);
+  }
+
+  ecma_number_t found_index = ECMA_NUMBER_MINUS_ONE;
+
+  int32_t increment = is_last_index_of ? -element_size : element_size;
+
+  for (int32_t position = (int32_t) from_index * element_size;
+       found_index < 0
+       && (is_last_index_of ? position >= 0 : (uint32_t) position < limit);
+       position += increment)
+  {
+    ecma_number_t element_num = ecma_get_typedarray_element (typedarray_buffer_p + position, class_id);
+    ecma_value_t element_value = ecma_make_number_value (element_num);
+
+    if (search_element == element_value)
+    {
+      found_index = (ecma_number_t) position / element_size;
+      break;
+    }
+
+    ecma_free_value (element_value);
+  }
+
+  if (ecma_is_value_empty (ret_value))
+  {
+    ret_value = ecma_make_number_value (found_index);
+  }
+
+  return ret_value;
```

Comment:
We could return directly from the loop if the element is found, otherwise we could always return with a `-1` value here. There would be no need for the `found_index` variable.